ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -1165,49 +1189,68 @@ withConnectionManager ConnectionManagerArguments {
+                       )
 
               -- the inbound protocol governor was supposed to call
               -- 'demotedToColdRemote' first.
-              InboundState connId connThread _handle dataFlow ->
-                assert False $ do
+              InboundState connId connThread _handle _dataFlow -> do
                 let connState' = TerminatingState connId connThread Nothing
                 writeTVar connVar connState'
                 return ( Just connThread
                        , Just (mkTransition connState connState')
-                       , UnsupportedState (InboundSt dataFlow) )
+                       , UnsupportedState st
+                       , Just (TrCMUnexpectedlyFalseAssertion
+                                 (UnregisterInboundConnection (Just connId)
+                                                              st)
+                                 )
+                       )
 
               -- the inbound connection governor ought to call
               -- 'demotedToColdRemote' first.
-              DuplexState connId connThread handle ->
-                assert False $ do
+              DuplexState connId connThread handle -> do
                 let connState' = OutboundDupState connId connThread handle Ticking
                 writeTVar connVar connState'
                 return ( Nothing
                        , Just (mkTransition connState connState')
-                       , UnsupportedState DuplexSt )
+                       , UnsupportedState st
+                       , Just (TrCMUnexpectedlyFalseAssertion
+                                 (UnregisterInboundConnection (Just connId)
+                                                              st)
+                                 )
+                       )
 
               -- If 'unregisterOutboundConnection' is called just before
               -- 'unregisterInboundConnection', the latter one might observe
               -- 'TerminatingState'.
               TerminatingState _connId _connThread _handleError ->
                 return ( Nothing
                        , Nothing
-                       , OperationSuccess CommitTr )
+                       , OperationSuccess CommitTr
+                       , Nothing
+                       )
               -- However, 'TerminatedState' should not be observable by
               -- 'unregisterInboundConnection', unless 'cmTimeWaitTimeout' is
               -- close to 'serverProtocolIdleTimeout'.
               TerminatedState _handleError ->
                 return ( Nothing
                        , Nothing
-                       , UnsupportedState TerminatedSt )
+                       , UnsupportedState TerminatedSt
+                       , Nothing
+                       )
 
       traverse_ (traceWith trTracer . TransitionTrace peerAddr) mbTransition
       traceCounters stateVar
 
       traverse_ cancel mbThread
-      return result
 
+      whenJust mbAssertion $ \tr' -> do
```

Comment:
is `tr` already in scope?

Reply:
I'll double check and change it accordingly!